Support DiscoveryConfig in terraform provider

avatus opened the pull request:
Closes https://github.com/gravitational/teleport/issues/60809

This will allow users to manage the DiscoveryConfig resource via terraform

Changelog: DiscoveryConfig resources can now be managed via the Teleport Terraform Provider

avatus committed:
Support DiscoveryConfig in terraform provider

Closes https://github.com/gravitational/teleport/issues/60809

This will allow users to manage the DiscoveryConfig resource via
terraform

GavinFrazar commented on integrations/terraform/protoc-gen-terraform-discoveryconfig.yaml:
```diff
@@ -0,0 +1,79 @@
+
+# These fields will be marked as Required: true
+required_fields:
+  - 'DiscoveryConfig.header.metadata'
```
You should add `DiscoveryConfig.header` as a required field as well, since it has required sub fields.

GavinFrazar commented on integrations/terraform/Makefile:
```diff
@@ -192,11 +200,7 @@ ifndef TERRAFORM_EXISTS
 	terraform -version
 	@exit -1
 endif
-# NOTE: This is related to an old bug in Terraform and will be fixed in future releases (possibly, by a workaround on our side)
```
This change doesn't look related to the PR - is it intended to be included?

GavinFrazar commented on lib/services/presets_test.go:
```diff
@@ -688,6 +688,7 @@ func TestAddRoleDefaults(t *testing.T) {
 									types.KindClusterNetworkingConfig,
 									types.KindDatabase,
 									types.KindDevice,
+									types.KindDiscoveryConfig,
```
I ran into this same test failure when I tried to fix this test for another resource lol
You just need to move this into the "expected" resource and put it in a separate rule as below, e.g.,
```
types.NewRule(types.KindDiscoveryConfig, RW()),
```

hugoShaka commented on docs/pages/reference/infrastructure-as-code/terraform-provider/data-sources/discovery_config.mdx:
```diff
@@ -0,0 +1,232 @@
+
+
+
+
```
The PR is missing an example in `integrations/terraform/examples`

hugoShaka committed:
Use TF types overrides instead of custom types

There no need to roll our own JoinMethod type, the join method
underlying type is a regular string. We can tell the schema geenrator to
use the native string type instead of having to write our own types
(which is error prone, we already messed up the bool option once).

This commit requires the protoc-gen-terraform v3.0.3 (with a schema
override fix, unreleased as of now).

protoc-gen-tfschema 3.0.2 -> 3.0.3

marcoandredinis commented on integrations/terraform/examples/resources/teleport_discovery_config/resource.tf:
```diff
@@ -0,0 +1,74 @@
+        join_method = "iam"
+        join_token  = "aws-discovery-token"
+        script_name = "default-installer"
+      }
```
There are a couple of values that are set when the `install_params` is not set in the `teleport.yaml/discovery_service.aws[].install` or in `discovery_config.spec.aws[].install`.


As an example, if the install_params is not set, it will set the:
- enroll_mode to `script` enum value
- install_teleport to `true`

I'm not sure we can do the same logic here, but if possible let's at least ensure the example value is set to those values.

rosstimothy commented on integrations/terraform/tfschema/discoveryconfig/v1/discoveryconfig_terraform.go:
Hrmm are you sure, the label is size/xl and the bot left a comment asking for an admin reviewer.
